airflow/models/deadlines.py

```diff
@@ -0,0 +1,100 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+from datetime import datetime
+from typing import TYPE_CHECKING
+
+import sqlalchemy_jsonfield
+from sqlalchemy import Column, DateTime, Index, Integer, String
+
+from airflow.models.base import Base, StringID
+from airflow.settings import json
+from airflow.utils.log.logging_mixin import LoggingMixin
+from airflow.utils.session import NEW_SESSION, provide_session
+
+if TYPE_CHECKING:
+    from sqlalchemy.orm import Session
+
+
+class Deadlines(LoggingMixin):
+    """A Deadline is a 'need-by' date which triggers a callback if the provided time has passed."""
+
+    @classmethod
+    @provide_session
+    def add_deadline(cls, deadline: DeadlinesModel, session: Session = NEW_SESSION):
+        """Add the provided deadline to the table."""
+        session.add(deadline)
+
+
+class DeadlinesModel(Base, LoggingMixin):
+    """Table containing Deadline Alert properties."""
+
+    __tablename__ = "deadlines"
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+
+    # If the Deadline Alert is for a DAG, store the DAG ID and Run ID from the dag_run.
+    """
+    TODO FIXME:   making these foreign keys causes the following error:
```

Comment:
I think the issue comes from the confusion between the columns `id` and `run_id` in `DagRun`. The foreign key should reference the primary key that is `id`, not `run_id`.

Reply:
Got it, I didn't realize an fkey HAD to be the pkey on the other table.  Thanks.

After some offline discussion, I'm going to drop this part entirely.  The thought was that it would make cleaning up easier.  If I had the fkey CASCADE, then the deadline would be removed when the run was, but someone raised the case where the dagrun may finish after the deadline but before the deadline callback fired.  In that case it should not be removed automatically when the dagrun finishes, because it did miss the deadline.  Removing the deadline will always - and only - be done intentionally.